File: site/indicators.js
// Mirrors scripts/indicators.py exactly, so manually-added tickers (computed
// live in the browser) score the same way as the daily-batch tracked ones.

function sma(values, period) {
  const out = new Array(values.length).fill(null);
  for (let i = 0; i < values.length; i++) {
    if (i + 1 < period) continue;
    const window = values.slice(i + 1 - period, i + 1);
    out[i] = window.reduce((a, b) => a + b, 0) / period;
  }
  return out;
}

function ema(values, period) {
  const out = new Array(values.length).fill(null);
  const k = 2 / (period + 1);
  let prev = null;
  for (let i = 0; i < values.length; i++) {
    if (prev === null) {
      if (i + 1 === period) {
        prev = values.slice(0, period).reduce((a, b) => a + b, 0) / period;
        out[i] = prev;
      }
      continue;
    }
    prev = values[i] * k + prev * (1 - k);
    out[i] = prev;
  }
  return out;
}

function rsi(closes, period = 14) {
  const out = new Array(closes.length).fill(null);
  const gains = [], losses = [];
  for (let i = 1; i < closes.length; i++) {
    const change = closes[i] - closes[i - 1];
    gains.push(Math.max(change, 0));
    losses.push(Math.max(-change, 0));
    if (i < period) continue;
    const windowGains = gains.slice(i - period, i);
    const windowLosses = losses.slice(i - period, i);
    const avgGain = windowGains.reduce((a, b) => a + b, 0) / period;
    const avgLoss = windowLosses.reduce((a, b) => a + b, 0) / period;
    out[i] = avgLoss === 0 ? 100 : 100 - 100 / (1 + avgGain / avgLoss);
  }
  return out;
}

function macd(closes, fast = 12, slow = 26, signal = 9) {
  const emaFast = ema(closes, fast);
  const emaSlow = ema(closes, slow);
  const macdLine = emaFast.map((a, i) => (a != null && emaSlow[i] != null ? a - emaSlow[i] : null));
  const valid = macdLine.filter(v => v != null);
  const signalValid = ema(valid, signal);
  const signalLine = new Array(macdLine.length - valid.length).fill(null).concat(signalValid);
  const histogram = macdLine.map((m, i) => (m != null && signalLine[i] != null ? m - signalLine[i] : null));
  return { macdLine, signalLine, histogram };
}

File: scripts/indicators.py
"""Pure-stdlib technical indicator helpers. No third-party dependencies on purpose,
so the daily GitHub Actions job needs nothing beyond the system Python."""


def sma(values, period):
    out = [None] * len(values)
    for i in range(len(values)):
        if i + 1 < period:
            continue
        window = values[i + 1 - period:i + 1]
        out[i] = sum(window) / period
    return out


def ema(values, period):
    out = [None] * len(values)
    k = 2 / (period + 1)
    prev = None
    for i, v in enumerate(values):
        if prev is None:
            if i + 1 == period:
                prev = sum(values[0:period]) / period
                out[i] = prev
            continue
        prev = v * k + prev * (1 - k)
        out[i] = prev
    return out


def rsi(closes, period=14):
    out = [None] * len(closes)
    gains, losses = [], []
    for i in range(1, len(closes)):
        change = closes[i] - closes[i - 1]
        gains.append(max(change, 0))
        losses.append(max(-change, 0))
        if i < period:
            continue
        window_gains = gains[i - period:i]
        window_losses = losses[i - period:i]
        avg_gain = sum(window_gains) / period
        avg_loss = sum(window_losses) / period
        if avg_loss == 0:
            out[i] = 100.0
        else:
            rs = avg_gain / avg_loss
            out[i] = 100 - (100 / (1 + rs))
    return out


def macd(closes, fast=12, slow=26, signal=9):
    ema_fast = ema(closes, fast)
    ema_slow = ema(closes, slow)
    macd_line = [
        (a - b) if (a is not None and b is not None) else None
        for a, b in zip(ema_fast, ema_slow)
    ]
    valid = [v for v in macd_line if v is not None]
    signal_valid = ema(valid, signal)
    signal_line = [None] * (len(macd_line) - len(valid)) + signal_valid
    histogram = [
        (m - s) if (m is not None and s is not None) else None
        for m, s in zip(macd_line, signal_line)
    ]
    return macd_line, signal_line, histogram
